frappe/model/document.py

```diff
@@ -89,6 +93,38 @@ def get_doc(*args, **kwargs):
 	raise ImportError(doctype)
 
 
+@contextmanager
+def read_only_document():
```

Comment:
where would you use it?

Reply:
I've implemented this [here](https://github.com/frappe/frappe/pull/26991/files#diff-5579cd03d81397823d516a6c3bf9cac1bfc1a7b6f1586de7447c410c319330e4R252) directly within the entry point to all conversion methods.

My goal is to declare a sub-scope in the call stack that is almost guaranteed to be read-only, except for rare and heavily discouraged direct database manipulations. This makes it significantly easier to reason about the code scope, enhancing maintainability, security, code quality, and conceptual separation of concerns.

For the conversion implementations in this PR, this approach ensures that stateful business logic (like insert and save) remains at the calling site. This makes the conversion implementations closer to pure functions, which are known for being more maintainable and secure over time.

An example of how this rigorous approach improved performance by reducing the call stack can be found [here](https://github.com/frappe/erpnext/pull/42160/files#diff-a7f7fdc2db51fe011675d0f9a7fb593cd56720070ddea9b391b6c3f816fe140eL528-R531).

Reply:
> I've implemented this [here](https://github.com/frappe/frappe/pull/26991/files#diff-5579cd03d81397823d516a6c3bf9cac1bfc1a7b6f1586de7447c410c319330e4R252) directly within the entry point to all conversion methods, because complex state on multiple doctypes is involved.

This is still within the same PR! It needs to be used as it should be in other parts

Reply:
> It needs to be used as it should be in other parts

Up to now, I have not thought of any other places where `read_only_document` could be useful.

Once this PR is agreed upon, it would make sense, in my opinion, to invest in a more comprehensive, yet gradual, refactoring initiative, but not necesarily using `read_only_document` directly, but by implementing doctype casters throughout the ERPNext codebase.

An even more strict variant `read_only_document_object` would be useful inside the `validate` hook to ensure separation of conern with the goal that all mutations are done in the `before_validate` hook as the only way to implement code to the effect of the already conventionalized `set_missing_values`.

In a similar way and more generally speaking, code quality might increase if certain sections of the lifecycle would exhibt a purpose build document api with limitations that make unintended use not only a breach of convention, but a hard code exception throught the ecosystem.